hugegraph-core/src/main/java/org/apache/hugegraph/util/collection/IntMapByDynamicHash.java

```diff
@@ -0,0 +1,794 @@
+                    if (toCopyEntry.getNext() == null) {
+                        newEntry = toCopyEntry; // no need to duplicate
+                    } else {
+                        newEntry = new Entry(toCopyEntry.getKey(), toCopyEntry.getValue());
+                    }
+                } else {
+                    newEntry =
+                        new Entry(toCopyEntry.getKey(), toCopyEntry.getValue(), o);
+                }
+                if (IntMapByDynamicHash.casTableAt(currentArray, index, o, newEntry)) {
+                    return;
+                }
+            }
+        }
+    }
+
+    private static final class ResizeContainer extends Entry {
+        private static final int QUEUE_INCREMENT =
+            Math.min(1 << 10,
+                     Integer.highestOneBit(Runtime.getRuntime().availableProcessors()) << 4);
+        private final AtomicInteger resizers = new AtomicInteger(1);
+        private final Entry[] nextArray;
+        private final AtomicInteger queuePosition;
+
+        private ResizeContainer(Entry[] nextArray, int oldSize) {
+            super(NULL_VALUE, NULL_VALUE, 4);
+            this.nextArray = nextArray;
+            this.queuePosition = new AtomicInteger(oldSize);
+        }
+
+        public void incrementResizer() {
+            this.resizers.incrementAndGet();
+        }
+
+        public void decrementResizerAndNotify() {
+            int remaining = this.resizers.decrementAndGet();
+            if (remaining == 0) {
+                synchronized (this) {
+                    this.notifyAll();
+                }
+            }
+        }
+
+        public int getQueuePosition() {
+            return this.queuePosition.get();
+        }
+
+        public int subtractAndGetQueuePosition() {
+            return this.queuePosition.addAndGet(-QUEUE_INCREMENT);
+        }
+
+        public void waitForAllResizers() {
+            if (this.resizers.get() > 0) {
+                for (int i = 0; i < 16; i++) {
+                    if (this.resizers.get() == 0) {
+                        break;
+                    }
+                }
+                for (int i = 0; i < 16; i++) {
+                    if (this.resizers.get() == 0) {
+                        break;
+                    }
+                    Thread.yield();
+                }
+            }
+            if (this.resizers.get() > 0) {
+                synchronized (this) {
+                    while (this.resizers.get() > 0) {
+                        try {
+                            this.wait();
+                        } catch (InterruptedException e) {
+                            //ginore
+                        }
+                    }
+                }
+            }
+        }
+
+        public boolean isNotDone() {
+            return this.resizers.get() > 0;
+        }
+
+        public void zeroOutQueuePosition() {
+            this.queuePosition.set(0);
+        }
+    }
+
+    private static class Entry {
```

Comment:
we expect a primitive type instead of a class, it's too many objects because each Entry will be generated an object